Guide: Securely store Micronaut Application secrets in Azure Key Vault

sbodvanski opened the pull request.

sdelamo committed:
guide: Micronaut Framework Dependency Injection Types (#1071)

sdelamo committed:
guide: Spring Boot to Micronaut Framework - Application class (#1075)

timyates committed:
bug: @features@ and @features-words@ macros not working (#1081)

In https://github.com/micronaut-projects/micronaut-guides/pull/1070 I had the regular
expression wrong for @features@ and @features-words@ expressions.

This fixes it so they are expanded correctly again

Closes #1073

sdelamo committed:
add twitter card

sdelamo committed:
twitter card

sdelamo committed:
method parameter not setter injection

dependabot[bot] committed:
Bump nimbus-jose-jwt from 9.21 to 9.25 in /buildSrc/src/main/resources (#1079)

Bumps [nimbus-jose-jwt](https://bitbucket.org/connect2id/nimbus-jose-jwt) from 9.21 to 9.25.
- [Changelog](https://bitbucket.org/connect2id/nimbus-jose-jwt/src/master/CHANGELOG.txt)
- [Commits](https://bitbucket.org/connect2id/nimbus-jose-jwt/branches/compare/9.25..9.21)

---
updated-dependencies:
- dependency-name: com.nimbusds:nimbus-jose-jwt
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

sdelamo committed:
explain how to run without test resources (#1089)

see: https://github.com/micronaut-projects/micronaut-guides/issues/1087

sdelamo committed:
add asciidoc indent logic (#1091)

sdelamo committed:
Guide executable JAR (#1090)

Commit:
Improve autonomous database Micronaut guide and fix several issues (#1088)

* Improve autonomous database Micronaut guide and fix several issues

* Improve autonomous database Micronaut guide
Fix review changes

sdelamo committed:
Data hibernate reactive kotlin version (#1094)

* feat: Micronaut Data Hibernate Reactive Guide

* refactor: Minor changes

* remove orElseThrow() it requires Java 10

Co-authored-by: roman <roman.naglic@gmail.com>

sdelamo committed:
Guide: How to use Micronaut Data from a Spring Boot app (#1080)

dependabot[bot] committed:
Bump micronaut-aws-sdk-v2 in /buildSrc/src/main/resources (#1096)

Bumps [micronaut-aws-sdk-v2](https://github.com/micronaut-projects/micronaut-aws) from 3.5.1 to 3.9.2.
- [Release notes](https://github.com/micronaut-projects/micronaut-aws/releases)
- [Commits](https://github.com/micronaut-projects/micronaut-aws/compare/v3.5.1...v3.9.2)

---
updated-dependencies:
- dependency-name: io.micronaut.aws:micronaut-aws-sdk-v2
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump s3 from 2.17.210 to 2.17.277 in /buildSrc/src/main/resources (#1095)

Bumps s3 from 2.17.210 to 2.17.277.

---
updated-dependencies:
- dependency-name: software.amazon.awssdk:s3
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

alvarosanchez committed:
Object Storage AWS guide (#1086)

sbodvanski committed:
Guide: Securely store Micronaut Application secrets in Azure Key Vault

graemerocher commented on guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc:
```diff
@@ -0,0 +1,194 @@
++   username: ${JDBC-USER}
++   password: ${JDBC-PASSWORD}
+    dialect: MYSQL
+    driverClassName: ${JDBC_DRIVER:com.mysql.cj.jdbc.Driver}
```
underscore is used here, is that correct?

sbodvanski replied:
Dashes are used because of the Azure naming convention that does not allow underscores. There is a note below this section in the Guide covering this Azure restriction.

`The Azure Key Vault secrets naming convention does not allow "_" to be used as a word joiner. Therefore, when adding JDBC secret placeholders, you should use "-" instead (e.q. JDBC-USER, JDBC-PASSWORD).`

sbodvanski replied:
<img width="738" alt="Underscore in Azure secret names" src="https://user-images.githubusercontent.com/4265345/191248299-44a80a0f-2a1d-40c5-b908-9a3dda488a3d.png">

graemerocher replied:
I am aware, but the reference above an underscore is used

sbodvanski replied:
Ah, I see your point.

sbodvanski replied:
Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/6a896fc232dd1a42e605a664cb7488b9ffbbe258.

graemerocher commented on guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc:
```diff
@@ -0,0 +1,194 @@
+image::azure-key-vault/azure-vault-create-vault-resouce.png[]
+{empty} +
+
+When Key Vault service deployment is ready, please go to the resource in order to create secrets. Prior to clicking on the Secrets option in the left menu list, please copy the Vault URI from the main screen section. You will need to set it as an environment variable in order to integrate Azure Key Vault into your application.
```
you can remove the use of the word "please" in this paragraph

sbodvanski replied:
Will do.

sbodvanski replied:
Covered by https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/8fed9a79fe5adb0859856581038fe14bef237f02.

graemerocher commented on guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc:
```diff
@@ -0,0 +1,194 @@
+In order to finish the Key Vault setup process. you will need to assign a proper policy to the Key Vault resource. Go to your Key Vault resource profile. Click on the Access policies option in the left menu. Finally, create an access policy.
+{empty} +
+
+image::azure-key-vault/azure-vault-create-policy.png[]
```
the policy should define the minimum amount of access. Is not just "List" and "Get" enough?

sbodvanski replied:
Yes, I think you are right. I will make the changes.

sbodvanski replied:
Covered by https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/8fed9a79fe5adb0859856581038fe14bef237f02.

sbodvanski committed:
Guide: Securely store Micronaut Application secrets in Azure Key Vault

Changes as per review comments.

sbodvanski committed:
Guide: Securely store Micronaut Application secrets in Azure Key Vault

alvarosanchez committed:
Switch to 3.7.0 (#1078)

* Fix compilation failures introduced by https://github.com/micronaut-projects/micronaut-starter/pull/1387 and https://github.com/micronaut-projects/micronaut-starter/pull/1404

* remove code using starter changes

* remove features already existing in starter

* annotate DynamoRepository as @Primary

* use @NonNull instad of @NotNull

* transactional falso for mongo async tests

* Micronaut 3.7.0

sdelamo committed:
no Nuts 🌰

sdelamo committed:
use version from the BOM

sdelamo committed:
Fix maven spring boot data guide (#1101)

* add missing dependencyManagement spring boot maven

* spring boot maven add annotation processors

sdelamo committed:
guide: Micronaut Framework Scopes (#1085)

sdelamo committed:
twitter card

[ci skip]

sdelamo commented on guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc:
```diff
@@ -0,0 +1,194 @@
+
+https://docs.micronaut.io/latest/guide/#bootstrap[Click here to get more details on distributed configuration]
+
+=== Clean up of the Application Configuration
```
```suggestion
=== Clean up the Application Configuration
```

sbodvanski replied:
Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

sdelamo commented on guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc:
```diff
@@ -0,0 +1,194 @@
+image::azure-key-vault/azure-vault-create-vault-resouce.png[]
+{empty} +
+
+When Key Vault service deployment is ready, go to the resource in order to create secrets. Prior to clicking on the Secrets option in the left menu list, copy the Vault URI from the main screen section. You will need to set it as an environment variable in order to integrate Azure Key Vault into your application.
```
```suggestion
When Key Vault service deployment is ready, go to the resource to create secrets. Before clicking on the Secrets option in the left menu list, copy the Vault URI from the main screen section. You will need to set it as an environment variable to integrate Azure Key Vault into your application.
```

sbodvanski replied:
Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

sdelamo commented on guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc:
```diff
@@ -0,0 +1,194 @@
+image::azure-key-vault/azure-vault-create-secret-jdbc-password.png[]
+{empty} +
+
+Go to App registration in order to register your application with the Microsoft identity platform. After registering your application,
```
```suggestion
Go to App registration to register your application with the Microsoft identity platform. After registering your application,
```

sbodvanski replied:
Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

sdelamo commented on guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc:
```diff
@@ -0,0 +1,194 @@
+{empty} +
+
+Go to App registration in order to register your application with the Microsoft identity platform. After registering your application,
+copy the secret Application (client) ID, and Directory (tenant) ID. You will use them later on as environment variables in order to connect to Azure Key Vault.
```
```suggestion
copy the secret Application (client) ID, and Directory (tenant) ID. You will use them later as environment variables to connect to Azure Key Vault.
```

sbodvanski replied:
Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

sdelamo commented on guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc:
```diff
@@ -0,0 +1,194 @@
+image::azure-key-vault/azure-register-application.png[]
+{empty} +
+
+Navigate to your application profile in order to generate application credentials that you will use to authenticate with the Azure Key Vault service. In the left menu, click on the Certificates & secrets. To generate the application credentials, click on the Client secrets tab and then on the New client secret button. Copy the secret value. You will need to set it along the Application (client) I ...
```
```suggestion
Navigate to your application profile to generate application credentials that you will use to authenticate with the Azure Key Vault service. In the left menu, click on the Certificates & secrets. To generate the application credentials, click on the Client secrets tab and then on the New client secret button. Copy the secret value. You must set it along the Application (client) ID, Directory (tenant) ID, and Vault URI as environment variables.
```

sbodvanski replied:
Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

sdelamo commented on guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc:
```diff
@@ -0,0 +1,194 @@
+image::azure-key-vault/azure-register-application-add-client-secret.png[]
+{empty} +
+
+In order to finish the Key Vault setup process. you will need to assign a proper policy to the Key Vault resource. Go to your Key Vault resource profile. Click on the Access policies option in the left menu. Select Get and List options from the Secret Management Operations list. Finally, create an access policy.
```
```suggestion
To finish the Key Vault setup process. You will need to assign a proper policy to the Key Vault resource. 

* Go to your Key Vault resource profile. 
* Click on the Access policies option in the left menu. 
* Select Get and List options from the Secret Management Operations list. 
* Finally, create an access policy.
```

sbodvanski replied:
Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

sdelamo commented on guides/micronaut-cloud-secrets-azure/micronaut-cloud-secrets-azure.adoc:
```diff
@@ -0,0 +1,194 @@
+
+With almost everything in place, you can start the application and try it out. First, set environment variables to configure the application datasource, then start the application.
+
+Create environment variables for `AZURE_CLIENT_ID`, `AZURE_TENANT_ID`,  `AZURE_SECRET_VALUE`, and `AZURE_VAULT_URI` which will be used in the Micronaut app's `application.yml` datasource:
```
```suggestion
Create environment variables for `AZURE_CLIENT_ID`, `AZURE_TENANT_ID`,  `AZURE_SECRET_VALUE`, and `AZURE_VAULT_URI`, which will be used in the Micronaut app's `application.yml` datasource:
```

sbodvanski replied:
Covered in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/27ffa927a580938d2c8edd451d078f4c25369a2f.

sdelamo reviewed (commented):
The screenshots' application name is `DemoApp`. Should it be `micronautguide` to match the value in `bootstrap.yml`?

sbodvanski committed:
Guide: Securely store Micronaut Application secrets in Azure Key Vault

Added changes as per review comments.

sdelamo committed:
extract common-distributed configuration

sdelamo committed:
use bootstrap.yml and feature of guide

sbodvanski committed:
Guide: Securely store Micronaut Application secrets in Azure Key Vault

Changed Azure registered app name to reflect the application name in micronaut configuration.

sbodvanski commented:
> The screenshots' application name is `DemoApp`. Should it be `micronautguide` to match the value in `bootstrap.yml`?

Changed in https://github.com/micronaut-projects/micronaut-guides/pull/1099/commits/3c3bafe47a4a8c1d66033dea8d7f5e0f53a621c0

graemerocher commented:
something went wrong with the merge?

sdelamo committed:
Merge branch 'main' into micronaut-cloud-secrets-azure

sdelamo committed:
Merge branch 'main' into improvements

sdelamo committed:
Merge branch 'improvements' into micronaut-cloud-secrets-azure

sdelamo committed:
use - not _

sdelamo committed:
add base

sdelamo committed:
fix features merge

sdelamo committed:
Revert "use - not _"

This reverts commit 13bb175dd03310f3c1d082482202091715722071.

sdelamo committed:
use invisibleFeatures instead of feature
